crates/nu-utils/standard_library/std.nu

```diff
@@ -1,85 +1,96 @@
-#    ·         ─────┬────
-#    ·              ╰── originates from here
-#  7 │     }
-#    ╰────
+#   × Assertion failed: 
+#     ╭─[myscript.nu:76:1]
+#  76 │ let a = 3
+#  77 │ assert eq $a "a string"
+#     ·           ──────┬──────
+#     ·                 ╰── Different types cannot be equal: int <-> string.
+#  78 │
+#     ╰────
+#
 #
-# >_ assert_eq $a 3
-# >_ assert_eq $a 1
+# >_ let a = 3
+# >_ assert eq $a 3
+# >_ assert eq $a 1
 # Error:
-#   × left is not equal to right
-#    ╭─[entry #12:5:1]
-#  5 │     if not $cond {
-#  6 │         error make {msg: $msg}
-#    ·         ─────┬────
-#    ·              ╰── originates from here
-#  7 │     }
-#    ╰────
+#   × Assertion failed: 
+#     ╭─[myscript.nu:81:1]
+#  81 │ assert eq $a 3
+#  82 │ assert eq $a 1
+#     ·           ──┬─
+#     ·             ╰── They are not equal: 3 != 1
+#  83 │
+#     ╰────
 # ```
-export def "assert eq" [left: any, right: any] {
-    _assert (($left | describe) == ($right | describe)) $"left and right operand of `assert eq` should have the same type"
-    _assert ($left == $right) "left is not equal to right"
+export def "assert eq" [left: any, right: any, message?: string] {
+    let left_type = ($left | describe)
+    let right_type = ($right | describe)
+    let left_start = (metadata $left).span.start
+    let right_end = (metadata $right).span.end
+
+    if ($left_type != $right_type) {
+        _assertion-error $left_start $right_end $"Different types cannot be equal: ($left_type) <-> ($right_type)." $message
+    }
+    if ($left != $right) {
+        _assertion-error $left_start $right_end $"They are not equal: ($left) != ($right)" $message
+    }
 }
 
 # ```nushell
-# >_ assert_ne $a "a string"
+# >_ let a = 3
+# >_ assert ne $a 1
+# >_ assert ne $a 3
 # Error:
-#   × left and right operand of `assert eq` should have the same type
-#    ╭─[entry #12:5:1]
-#  5 │     if not $cond {
-#  6 │         error make {msg: $msg}
-#    ·         ─────┬────
-#    ·              ╰── originates from here
-#  7 │     }
-#    ╰────
-#
-# >_ assert_ne $a 1
-# >_ assert_ne $a 3
-# Error:
-#   × left is equal to right
-#    ╭─[entry #12:5:1]
-#  5 │     if not $cond {
-#  6 │         error make {msg: $msg}
-#    ·         ─────┬────
-#    ·              ╰── originates from here
-#  7 │     }
-#    ╰────
+#   × Assertion failed:
+#      ╭─[C:\Users\fm\git\nushell\crates\nu-utils\standard_library\std.nu:113:1]
+#  113 │ assert ne $a 1
+#  114 │ assert ne $a 3
+#      ·           ──┬─
+#      ·             ╰── They both are 3
+#  115 │
+#      ╰────
 # ```
-export def "assert ne" [left: any, right: any] {
-    _assert (($left | describe) == ($right | describe)) $"left and right operand of `assert eq` should have the same type"
-    _assert ($left != $right) "left is equal to right"
+export def "assert ne" [left: any, right: any, message?: string] {
+    let left_start = (metadata $left).span.start
+    let right_end = (metadata $right).span.end
+
```

Comment:
you're just missing the type check here :+1: 
just to have `assert eq` and `assert ne` be consistent between each other :yum:

Reply:
Sorry, I just forgot to update the tests, it was my mistake.

I also want consistency, see https://github.com/nushell/nushell/pull/8405#issuecomment-1464944357. If `$a == $b` then `assert eq $a $b` should pass. Similarly if `$a != $b` then `assert ne $a $b` should pass. I do not agree that `assert ne 1 "2"` should fail.

I can update the tests to assert to the right behaviour or revert changing `assert ne`, then  it is up to you if you want implement the new style for it as well. (Just please document that it is intentionally not consistent with the `!=` operator.)

Reply:
I updated the tests.